lsp/nls/src/requests/completion.rs

```diff
@@ -29,18 +31,117 @@ use crate::{
+                })
+            })
+            .unwrap_or_else(|| self.ty.to_string())
+    }
+
+    fn completion_item_kind(&self) -> CompletionItemKind {
+        match &self.ty {
+            ty if ty.is_function_type() => CompletionItemKind::Function,
+            _ => CompletionItemKind::Property,
+        }
+    }
+
+    fn to_lsp_completion_item(&self) -> CompletionItem {
+        /// Attach quotes to a non-ASCII string
+        fn adjust_name(name: &str) -> String {
+            if name.is_ascii() {
+                String::from(name)
+            } else {
+                format!("\"{}\"", name)
+            }
+        }
+        let doc = || {
+            let item = self.item.as_ref()?;
+            let meta = item.meta.as_ref()?;
+            let doc = meta.doc.as_ref()?;
+            let doc = Documentation::MarkupContent(MarkupContent {
+                kind: MarkupKind::Markdown,
+                value: format!("{}\nMerge Priority {}", doc.clone(), meta.priority),
+            });
+            Some(doc)
+        };
+        CompletionItem {
+            label: adjust_name(self.ident.label()),
+            detail: Some(self.detail()),
+            kind: Some(self.completion_item_kind()),
+            documentation: doc(),
+            ..Default::default()
+        }
+    }
+}
+
 /// Find the record field associated with a particular ID in the linearization
 /// using lexical scoping rules.
 fn find_fields_from_term_kind(
     linearization: &Completed,
     id: ItemId,
     path: &mut Vec<Ident>,
-) -> Option<Vec<Ident>> {
+) -> Option<Vec<IdentWithType>> {
     let item = linearization.get_item(id)?;
     match item.kind {
         TermKind::Record(ref fields) => {
             if path.is_empty() {
-                Some(fields.keys().cloned().collect())
+                Some(
+                    fields
+                        .keys()
+                        .cloned()
+                        .map(|ident| {
+                            // This unwrap is safe because this `ident` is gotten from `fields`
+                            // which we're currently iterating over. So we're sure `ident` is a
+                            // is present in `fields`
+                            let id = fields.get(&ident).unwrap();
+                            // This unwrap is safe because, `id` is the field of the record
+                            // we're currently analyzing. We're sure that the linearization
+                            // phase doesn't produce wrong or invalid ids.
+                            let item = linearization.get_item(*id).unwrap();
+                            let (ty, _) = linearization.resolve_item_type_meta(item);
+                            IdentWithType {
+                                ident,
+                                ty,
+                                item: Some(item.clone()),
+                            }
+                        })
```

Comment:
I think you could just iterate over pairs of `(key, value)` using `keys.iter()` instead of having to use `field.get(&ident).unwrap()`

Reply:
Yes, thanks!